core/src/main/java/org/apache/iceberg/avro/PruneColumns.java

```diff
@@ -345,4 +345,27 @@ private static boolean isOptionSchemaWithNonNullFirstOption(Schema schema) {
     return AvroSchemaUtil.isOptionSchema(schema)
         && schema.getTypes().get(0).getType() != Schema.Type.NULL;
   }
+
+  // for primitive types, the visitResult will be null, we want to reuse the primitive types from
+  // the original
+  // schema, while for nested types, we want to use the visitResult because they have content from
+  // the previous
+  // recursive calls.
+  private static Schema copyUnion(Schema record, List<Schema> visitResults) {
+    List<Schema> branches = Lists.newArrayListWithExpectedSize(visitResults.size());
+    for (int i = 0; i < visitResults.size(); i++) {
+      if (visitResults.get(i) == null) {
+        branches.add(record.getTypes().get(i));
+      } else {
+        branches.add(visitResults.get(i));
+      }
+    }
+    Schema schema = Schema.createUnion(branches);
+    if (record.getObjectProp(SchemaToType.AVRO_FIELD_NAME_TO_ICEBERG_ID) != null) {
```

Comment:
@yiqiangin, @wmoustafa, I would expect this to apply the name mapping, but instead it passes on a custom schema property. I don't think that this approach is correct.

The Avro implementation for name mapping is a little odd. For Parquet and ORC, there's a class that rewrites the schema and adds IDs. It looks like instead of taking that approach, the Avro implementer added name mapping to this class. That's okay, but that means that the name mapping should be applied here for the union work.

We want to create guarantees that we can rely on to simplify other code. In this case, once `PruneColumns` is done, we're guaranteed to have an Avro schema with the correct field IDs annotated throughout.

To do that, I think the field ID should be added to each Schema that is a branch of the union:

```java
  List<Schema> unionTypes = union.getTypes();
  for (int ind = 0; ind < branches.size(); ind += 1) {
    Schema branchSchema = visitResults.get(ind);
    if (branchSchema == null) {
      branchSchema = unionTypes.get(ind);
    }

    Integer branchId = AvroSchemaUtil.getBranchId(branchSchema, nameMapping, fieldNames());
    if (branchId != null) {
      optionSchema.addProp(AvroSchemaUtil.BRANCH_ID_PROP, String.valueOf(branchId));
    }

    branches.add(branchSchema)
  }

  return Schema.createUnion(branches);

// AvroSchemaUtil additions:

  public static final String BRANCH_ID_PROP = "branch-id";

  static Integer getBranchId(
      Schema branch, NameMapping mapping, Iterable<String> parentFieldNames) {
    Object id = branch.getObjectProp(BRANCH_ID_PROP);
    if (id != null) {
      return toInt(id);
    } else if (mapping != null) {
      MappedField mappedField = findInMapping(mapping, parentFieldNames, branch.getName(), branch.getFullName());
      if (mappedField != null) {
        return mappedField.id();
      }
    }

    return null;
  }

  private static MappedField findInMapping(NameMapping mapping, Iterable<String> parentFieldNames, String... nameOpts) {
    List<String> names = Lists.newArrayList(parentFieldNames);
    for (String name : nameOpts) {
      names.add(name);
      MappedField field = mapping.find(name);
      if (field != null) {
        return field;
      }
    }

    return null;
  }
```